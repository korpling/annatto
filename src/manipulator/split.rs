use std::{collections::BTreeMap, ops::Not};

use documented::{Documented, DocumentedFields};
use graphannis::{
    graph::AnnoKey,
    update::{GraphUpdate, UpdateEvent},
};
use graphannis_core::{annostorage::ValueSearch, graph::NODE_NAME_KEY};
use itertools::Itertools;
use serde::Deserialize;
use struct_field_names_as_array::FieldNamesAsSlice;

use crate::{
    core::update_graph, deserialize::deserialize_anno_key, error::Result,
    progress::ProgressReporter,
};

use super::Manipulator;

/// This operation splits conflated annotation values into individual annotations.
#[derive(Deserialize, Documented, DocumentedFields, FieldNamesAsSlice)]
#[serde(deny_unknown_fields)]
pub struct SplitValues {
    /// This is the delimiter between the parts of the conflated annotation in the input graph
    #[serde(default = "default_delimiter")]
    delimiter: String,
    /// The annotation that holds the conflated values.
    #[serde(deserialize_with = "deserialize_anno_key")]
    anno: AnnoKey,
    /// This maps a target annotation name to a list of potential values to be found in the split parts.
    #[serde(default)]
    layers: Vec<Layer>,
    /// If set to `true`, the original annotations will be deleted.
    #[serde(default)]
    delete: bool,
}

#[derive(Deserialize)]
#[serde(untagged)]
enum Layer {
    ByIndex {
        index: usize,
        #[serde(deserialize_with = "deserialize_anno_key")]
        key: AnnoKey,
    },
    ByValues {
        #[serde(deserialize_with = "deserialize_anno_key")]
        key: AnnoKey,
        #[serde(default)]
        values: Vec<String>,
    },
}

const DEFAULT_DELIMITER: &str = "-";

fn default_delimiter() -> String {
    DEFAULT_DELIMITER.to_string()
}

impl Manipulator for SplitValues {
    fn manipulate_corpus(
        &self,
        graph: &mut graphannis::AnnotationGraph,
        _workflow_directory: &std::path::Path,
        step_id: crate::StepID,
        tx: Option<crate::workflow::StatusSender>,
    ) -> std::result::Result<(), Box<dyn std::error::Error>> {
        let mut update = GraphUpdate::default();
        let node_annos = graph.get_node_annos();
        let (ns, name) = {
            (
                self.anno.ns.is_empty().not().then(|| self.anno.ns.as_str()),
                self.anno.name.as_str(),
            )
        };
        let value_map = self
            .layers
            .iter()
            .flat_map(|l| match l {
                Layer::ByIndex { .. } => vec![],
                Layer::ByValues { key, values } => {
                    values.iter().map(|vv| (vv.as_str(), key)).collect_vec()
                }
            })
            .collect::<BTreeMap<&str, &AnnoKey>>();
        let index_to_key: BTreeMap<usize, &AnnoKey> = self
            .layers
            .iter()
            .filter_map(|l| match l {
                Layer::ByIndex { index, key } => Some((*index, key)),
                Layer::ByValues { .. } => None,
            })
            .collect();
        let match_vec = node_annos
            .exact_anno_search(ns, name, ValueSearch::Any)
            .collect_vec();
        let progress = ProgressReporter::new(tx.clone(), step_id.clone(), match_vec.len())?;
        let anno_as_key = AnnoKey {
            name: name.into(),
            ns: ns.unwrap_or("").into(),
        };
        for m in match_vec {
            let n = m?.node;
            if let Some(node_name) = node_annos.get_value_for_item(&n, &NODE_NAME_KEY)? {
                if self.delete {
                    update.add_event(UpdateEvent::DeleteNodeLabel {
                        node_name: node_name.to_string(),
                        anno_ns: ns.unwrap_or("").to_string(),
                        anno_name: name.to_string(),
                    })?;
                }
                if let Some(v) = node_annos.get_value_for_item(&n, &anno_as_key)? {
                    self.map(&mut update, &node_name, &v, &value_map, &index_to_key)?;
                };
            }
            progress.worked(1)?;
        }
        update_graph(graph, &mut update, Some(step_id), tx)?;
        Ok(())
    }

    fn requires_statistics(&self) -> bool {
        false
    }
}

impl SplitValues {
    fn map(
        &self,
        update: &mut GraphUpdate,
        node_name: &str,
        value: &str,
        value_map: &BTreeMap<&str, &AnnoKey>,
        index_map: &BTreeMap<usize, &AnnoKey>,
    ) -> Result<()> {
        for (i, v) in value.split(&self.delimiter).enumerate() {
            if let Some(key) = index_map.get(&(i + 1)) {
                update.add_event(UpdateEvent::AddNodeLabel {
                    node_name: node_name.to_string(),
                    anno_ns: key.ns.to_string(),
                    anno_name: key.name.to_string(),
                    anno_value: v.to_string(),
                })?;
            } else {
                // use value map
                if let Some(anno_key) = value_map.get(v) {
                    update.add_event(UpdateEvent::AddNodeLabel {
                        node_name: node_name.to_string(),
                        anno_ns: anno_key.ns.to_string(),
                        anno_name: anno_key.name.to_string(),
                        anno_value: v.to_string(),
                    })?;
                }
            }
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::{fs, path::Path};

    use graphannis::AnnotationGraph;
    use insta::assert_snapshot;

    use crate::{
        exporter::graphml::GraphMLExporter,
        importer::{treetagger::ImportTreeTagger, Importer},
        manipulator::{split::SplitValues, Manipulator},
        test_util::export_to_string,
    };

    #[test]
    fn split_features() {
        let importer = ImportTreeTagger::default();
        let input_path = Path::new("tests/data/graph_op/split_values/");
        let u = importer.import_corpus(
            input_path,
            crate::StepID {
                module_name: "import_treetagger".to_string(),
                path: None,
            },
            None,
        );
        assert!(u.is_ok());
        let mut update = u.unwrap();
        let g = AnnotationGraph::with_default_graphstorages(true);
        assert!(g.is_ok());
        let mut graph = g.unwrap();
        assert!(graph.apply_update(&mut update, |_| {}).is_ok());
        let config_path = "tests/data/graph_op/split_values/config.toml";
        let tstr = fs::read_to_string(config_path);
        assert!(tstr.is_ok());
        let toml_str = tstr.unwrap();
        let op: Result<SplitValues, _> = toml::from_str(toml_str.as_str());
        assert!(
            op.is_ok(),
            "An error occured on deserialization: {:?}",
            op.err()
        );
        let split_op = op.unwrap();
        assert!(split_op
            .manipulate_corpus(
                &mut graph,
                Path::new("./"),
                crate::StepID {
                    module_name: "split_conflated".to_string(),
                    path: None
                },
                None
            )
            .is_ok());
        let gml = export_to_string(&graph, GraphMLExporter::default());
        assert!(gml.is_ok());
        let graphml = gml.unwrap();
        let start = graphml.find("<node ").unwrap_or(0);
        let end = graphml.rfind("</node>").unwrap_or(usize::MAX);
        assert_snapshot!(&graphml[start..end + 7]);
    }
}
