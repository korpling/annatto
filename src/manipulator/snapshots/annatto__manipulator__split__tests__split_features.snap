---
source: src/manipulator/split.rs
expression: graphml
---
<?xml version="1.0" encoding="UTF-8"?>
<graphml>
    <key id="k0" for="graph" attr.name="configuration" attr.type="string"/>
    <key id="k1" for="node" attr.name="Case" attr.type="string"/>
    <key id="k2" for="node" attr.name="Gender" attr.type="string"/>
    <key id="k3" for="node" attr.name="Mood" attr.type="string"/>
    <key id="k4" for="node" attr.name="Number" attr.type="string"/>
    <key id="k5" for="node" attr.name="Person" attr.type="string"/>
    <key id="k6" for="node" attr.name="Tense" attr.type="string"/>
    <key id="k7" for="node" attr.name="VerbClass" attr.type="string"/>
    <key id="k8" for="node" attr.name="annis::doc" attr.type="string"/>
    <key id="k9" for="node" attr.name="annis::layer" attr.type="string"/>
    <key id="k10" for="node" attr.name="default_ns::lemma" attr.type="string"/>
    <key id="k11" for="node" attr.name="annis::node_type" attr.type="string"/>
    <key id="k12" for="node" attr.name="default_ns::pos" attr.type="string"/>
    <key id="k13" for="node" attr.name="annis::tok" attr.type="string"/>
    <key id="k14" for="node" attr.name="annis::tok-whitespace-before" attr.type="string"/>
    <graph edgedefault="directed" parse.order="nodesfirst" parse.nodeids="free" parse.edgeids="canonical">
        <data key="k0"><![CDATA[
# configure visualizations here
]]></data>
        <node id="split_values">
            <data key="k11">corpus</data>
        </node>
        <node id="split_values/test_file">
            <data key="k11">corpus</data>
            <data key="k8">test_file</data>
        </node>
        <node id="split_values/test_file#text">
            <data key="k11">datasource</data>
        </node>
        <node id="split_values/test_file#t1">
            <data key="k1">Nom</data>
            <data key="k10">der</data>
            <data key="k11">node</data>
            <data key="k12">DDART</data>
            <data key="k13">die</data>
            <data key="k2">Fem</data>
            <data key="k4">Sg</data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t2">
            <data key="k1">Nom</data>
            <data key="k10">zeit</data>
            <data key="k11">node</data>
            <data key="k12">NA</data>
            <data key="k13">zeit</data>
            <data key="k14"> </data>
            <data key="k2">Fem</data>
            <data key="k4">Sg</data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t3">
            <data key="k1">Gen</data>
            <data key="k10">der</data>
            <data key="k11">node</data>
            <data key="k12">DDART</data>
            <data key="k13">des</data>
            <data key="k14"> </data>
            <data key="k4">Sg</data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t4">
            <data key="k1">Gen</data>
            <data key="k10">schweigen</data>
            <data key="k11">node</data>
            <data key="k12">NA</data>
            <data key="k13">schweygens</data>
            <data key="k14"> </data>
            <data key="k4">Sg</data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t5">
            <data key="k10">sein</data>
            <data key="k11">node</data>
            <data key="k12">VAFIN</data>
            <data key="k13">ist</data>
            <data key="k14"> </data>
            <data key="k3">Ind</data>
            <data key="k4">Sg</data>
            <data key="k5">3</data>
            <data key="k6">Präs</data>
            <data key="k7">Unr</data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t6">
            <data key="k10">vor|vorgehen</data>
            <data key="k11">node</data>
            <data key="k12">PTKVZ|VVPP</data>
            <data key="k13">vorgangen</data>
            <data key="k14"> </data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t7">
            <data key="k10">,</data>
            <data key="k11">node</data>
            <data key="k12">$_</data>
            <data key="k13">,</data>
            <data key="k14"> </data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t8">
            <data key="k10">und</data>
            <data key="k11">node</data>
            <data key="k12">KON</data>
            <data key="k13">und</data>
            <data key="k14"> </data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t9">
            <data key="k1">Akk</data>
            <data key="k10">der</data>
            <data key="k11">node</data>
            <data key="k12">DDART</data>
            <data key="k13">die</data>
            <data key="k14"> </data>
            <data key="k2">Fem</data>
            <data key="k4">Sg</data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t10">
            <data key="k1">Akk</data>
            <data key="k10">zeit</data>
            <data key="k11">node</data>
            <data key="k12">NA</data>
            <data key="k13">zeit</data>
            <data key="k14"> </data>
            <data key="k2">Fem</data>
            <data key="k4">Sg</data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t11">
            <data key="k10">zu</data>
            <data key="k11">node</data>
            <data key="k12">PTKZU</data>
            <data key="k13">zu</data>
            <data key="k14"> </data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t12">
            <data key="k10">reden</data>
            <data key="k11">node</data>
            <data key="k12">VVINF</data>
            <data key="k13">reden</data>
            <data key="k14"> </data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t13">
            <data key="k10">sein</data>
            <data key="k11">node</data>
            <data key="k12">VAFIN</data>
            <data key="k13">ist</data>
            <data key="k14"> </data>
            <data key="k3">Ind</data>
            <data key="k4">Sg</data>
            <data key="k5">3</data>
            <data key="k6">Präs</data>
            <data key="k7">Unr</data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t14">
            <data key="k10">kommen</data>
            <data key="k11">node</data>
            <data key="k12">VVPP</data>
            <data key="k13">kommen</data>
            <data key="k14"> </data>
            <data key="k9">default_layer</data>
        </node>
        <node id="split_values/test_file#t15">
            <data key="k10">,</data>
            <data key="k11">node</data>
            <data key="k12">$_</data>
            <data key="k13">,</data>
            <data key="k14"> </data>
            <data key="k9">default_layer</data>
        </node>
        <edge id="e0" source="split_values/test_file#t14" target="split_values/test_file#t15" label="Ordering/annis/">
        </edge>
        <edge id="e1" source="split_values/test_file#t13" target="split_values/test_file#t14" label="Ordering/annis/">
        </edge>
        <edge id="e2" source="split_values/test_file#t12" target="split_values/test_file#t13" label="Ordering/annis/">
        </edge>
        <edge id="e3" source="split_values/test_file#t11" target="split_values/test_file#t12" label="Ordering/annis/">
        </edge>
        <edge id="e4" source="split_values/test_file#t10" target="split_values/test_file#t11" label="Ordering/annis/">
        </edge>
        <edge id="e5" source="split_values/test_file#t9" target="split_values/test_file#t10" label="Ordering/annis/">
        </edge>
        <edge id="e6" source="split_values/test_file#t8" target="split_values/test_file#t9" label="Ordering/annis/">
        </edge>
        <edge id="e7" source="split_values/test_file#t7" target="split_values/test_file#t8" label="Ordering/annis/">
        </edge>
        <edge id="e8" source="split_values/test_file#t6" target="split_values/test_file#t7" label="Ordering/annis/">
        </edge>
        <edge id="e9" source="split_values/test_file#t5" target="split_values/test_file#t6" label="Ordering/annis/">
        </edge>
        <edge id="e10" source="split_values/test_file#t4" target="split_values/test_file#t5" label="Ordering/annis/">
        </edge>
        <edge id="e11" source="split_values/test_file#t3" target="split_values/test_file#t4" label="Ordering/annis/">
        </edge>
        <edge id="e12" source="split_values/test_file#t2" target="split_values/test_file#t3" label="Ordering/annis/">
        </edge>
        <edge id="e13" source="split_values/test_file#t1" target="split_values/test_file#t2" label="Ordering/annis/">
        </edge>
        <edge id="e14" source="split_values/test_file" target="split_values" label="PartOf/annis/">
        </edge>
        <edge id="e15" source="split_values/test_file#text" target="split_values/test_file" label="PartOf/annis/">
        </edge>
        <edge id="e16" source="split_values/test_file#t1" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e17" source="split_values/test_file#t2" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e18" source="split_values/test_file#t3" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e19" source="split_values/test_file#t4" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e20" source="split_values/test_file#t5" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e21" source="split_values/test_file#t6" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e22" source="split_values/test_file#t7" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e23" source="split_values/test_file#t8" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e24" source="split_values/test_file#t9" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e25" source="split_values/test_file#t10" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e26" source="split_values/test_file#t11" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e27" source="split_values/test_file#t12" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e28" source="split_values/test_file#t13" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e29" source="split_values/test_file#t14" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
        <edge id="e30" source="split_values/test_file#t15" target="split_values/test_file#text" label="PartOf/annis/">
        </edge>
    </graph>
</graphml>
